feat(emoji): add string emoji property to reactions - DB and protocol

jrainville opened the pull request:
Fixes https://github.com/status-im/status-desktop/issues/18801

Adds a string field called `emoji` in the emoji_reactions table and in the protobuf that enable sending and receiving any emoji reaction.

Modifies the tests to show that it works and that it's backwards compatible

edit: added a second commit with the API changes too, since they are pretty small

jrainville committed:
feat(emoji): add string emoji property to reactions - DB and protocol

Fixes https://github.com/status-im/status-desktop/issues/18801

Adds a string field called `emoji` in the emoji_reactions table and in the protobuf that enable sending and receiving any emoji reaction.

Modifies the tests to show that it works and that it's backwards compatible

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (25)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | c9738d56 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6916/1/) | 2025-09-09 17:50:59 | ~2 min | `linux/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-174811-c9738d-pr6916.zip) |
| :heavy_check_mark: | c9738d56 | [#1](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6916/1/) | 2025-09-09 17:51:31 | ~3 min | `macos/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-174815-c9738d-pr6916.zip) |
| :heavy_check_mark: | c9738d56 | [#1](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6916/1/) | 2025-09-09 17:55:21 | ~6 min | `windows/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-174821-c9738d-pr6916.zip) |
| :heavy_check_mark: | c9738d56 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/nwaku/job/PR-6916/1/) | 2025-09-09 17:58:39 | ~10 min | `linux/nwaku` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/nwaku-250909-174811-c9738d-pr6916.zip) |
| :heavy_check_mark: | c9738d56 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/1/) | 2025-09-09 17:58:53 | ~10 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/1//consoleText) |
| :heavy_check_mark: | c9738d56 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/1/) | 2025-09-09 18:13:10 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | b6d057f9 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6916/2/) | 2025-09-09 17:53:50 | ~2 min | `linux/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-175106-b6d057-pr6916.zip) |
| :heavy_check_mark: | b6d057f9 | [#2](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6916/2/) | 2025-09-09 17:54:30 | ~2 min | `macos/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-175136-b6d057-pr6916.zip) |
| :heavy_check_mark: | b6d057f9 | [#2](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6916/2/) | 2025-09-09 18:02:09 | ~6 min | `windows/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-175525-b6d057-pr6916.zip) |
| :heavy_check_mark: | b6d057f9 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/2/) | 2025-09-09 18:09:27 | ~10 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/2//consoleText) |
| :heavy_check_mark: | b6d057f9 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/nwaku/job/PR-6916/2/) | 2025-09-09 18:09:37 | ~10 min | `linux/nwaku` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/nwaku-250909-175843-b6d057-pr6916.zip) |
| :heavy_check_mark: | b6d057f9 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/2/) | 2025-09-09 18:37:02 | ~23 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 45ac81f0 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6916/3/) | 2025-09-09 18:56:47 | ~2 min | `linux/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-185354-45ac81-pr6916.zip) |
| :heavy_check_mark: | 45ac81f0 | [#3](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6916/3/) | 2025-09-09 18:57:08 | ~3 min | `macos/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-185359-45ac81-pr6916.zip) |
| :heavy_check_mark: | 45ac81f0 | [#3](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6916/3/) | 2025-09-09 19:00:38 | ~6 min | `windows/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-185404-45ac81-pr6916.zip) |
| :heavy_check_mark: | 45ac81f0 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/3/) | 2025-09-09 19:02:59 | ~8 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/3//consoleText) |
| :heavy_check_mark: | 45ac81f0 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/nwaku/job/PR-6916/3/) | 2025-09-09 19:03:33 | ~9 min | `linux/nwaku` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/nwaku-250909-185356-45ac81-pr6916.zip) |
| :heavy_check_mark: | 45ac81f0 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/3/) | 2025-09-09 19:18:34 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 952c8e1f | [#4](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6916/4/) | 2025-09-09 20:13:21 | ~3 min | `macos/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-201012-952c8e-pr6916.zip) |
| :heavy_check_mark: | 952c8e1f | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6916/4/) | 2025-09-09 20:13:33 | ~3 min | `linux/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-201006-952c8e-pr6916.zip) |
| :heavy_check_mark: | 952c8e1f | [#4](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6916/4/) | 2025-09-09 20:16:54 | ~6 min | `windows/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250909-201019-952c8e-pr6916.zip) |
| :heavy_multiplication_x: | 952c8e1f | [#4](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/4/) | 2025-09-09 20:18:47 | ~8 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/4//consoleText) |
| :heavy_check_mark: | 952c8e1f | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/nwaku/job/PR-6916/4/) | 2025-09-09 20:20:18 | ~10 min | `linux/nwaku` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/nwaku-250909-201007-952c8e-pr6916.zip) |
| :heavy_check_mark: | 952c8e1f | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/4/) | 2025-09-09 20:34:29 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/4//consoleText) |
| :heavy_check_mark: | 952c8e1f | [#5](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/5/) | 2025-09-09 20:36:05 | ~7 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/5//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 6f6d11e1 | [#5](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6916/5/) | 2025-09-10 15:57:51 | ~3 min | `macos/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250910-155438-6f6d11-pr6916.zip) |
| :heavy_check_mark: | 6f6d11e1 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6916/5/) | 2025-09-10 16:01:18 | ~6 min | `linux/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250910-155433-6f6d11-pr6916.zip) |
| :heavy_check_mark: | 6f6d11e1 | [#5](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6916/5/) | 2025-09-10 16:01:39 | ~6 min | `windows/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250910-155443-6f6d11-pr6916.zip) |
| :heavy_check_mark: | 6f6d11e1 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/nwaku/job/PR-6916/5/) | 2025-09-10 16:10:04 | ~15 min | `linux/nwaku` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/nwaku-250910-155434-6f6d11-pr6916.zip) |
| :heavy_check_mark: | 6f6d11e1 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/6/) | 2025-09-10 16:12:05 | ~17 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/6//consoleText) |
| :heavy_check_mark: | 6f6d11e1 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/5/) | 2025-09-10 16:27:18 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 5cfcaa62 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6916/6/) | 2025-09-10 18:54:15 | ~2 min | `linux/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250910-185117-5cfcaa-pr6916.zip) |
| :heavy_check_mark: | 5cfcaa62 | [#6](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6916/6/) | 2025-09-10 18:55:00 | ~3 min | `macos/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250910-185121-5cfcaa-pr6916.zip) |
| :heavy_check_mark: | 5cfcaa62 | [#6](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6916/6/) | 2025-09-10 18:59:06 | ~7 min | `windows/status-go` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-250910-185126-5cfcaa-pr6916.zip) |
| :heavy_check_mark: | 5cfcaa62 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/7/) | 2025-09-10 19:00:30 | ~9 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6916/7//consoleText) |
| :heavy_check_mark: | 5cfcaa62 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/nwaku/job/PR-6916/6/) | 2025-09-10 19:01:04 | ~9 min | `linux/nwaku` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/nwaku-250910-185118-5cfcaa-pr6916.zip) |
| :heavy_check_mark: | 5cfcaa62 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/6/) | 2025-09-10 19:15:30 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6916/6//consoleText) |

jrainville committed:
feat(emoji): add API to send custom emojis

Fixes https://github.com/status-im/status-desktop/issues/18800

Adds a new API function that enables passing any emoji. Validates the the emoji is really an emoji too

igor-sirotin commented on protocol/protobuf/emoji_reaction.proto:
```diff
@@ -20,7 +20,7 @@ message EmojiReaction {
-  Type type = 5;
+  Type type = 5 [deprecated = true];
 
   enum Type {
```
We gotta deprecate `Type` as well I guess
```suggestion
  enum Type {
      option deprecated = true
```

igor-sirotin commented on protocol/emoji_reaction.go:
```diff
@@ -32,9 +32,13 @@ func NewEmojiReaction() *EmojiReaction {
+	if e.Emoji == "" {
+		return types.EncodeHex(crypto.Keccak256([]byte(fmt.Sprintf("%s%s%d", e.From, e.MessageId, e.Type))))
+	}
+	return types.EncodeHex(crypto.Keccak256([]byte(fmt.Sprintf("%s%s%s", e.From, e.MessageId, e.Emoji))))
```
```suggestion
    var data string
	if e.Emoji == "" {
	    data = fmt.Sprintf("%s%s%d", e.From, e.MessageId, e.Type)
	} else {
	    data = fmt.Sprintf("%s%s%s", e.From, e.MessageId, e.Emoji)
	}
	return types.EncodeHex(crypto.Keccak256([]byte(data))))
```

igor-sirotin commented on protocol/messenger_handler.go:
```diff
@@ -2901,6 +2901,11 @@ func (m *Messenger) HandleEmojiReaction(state *ReceivedMessageState, pbEmojiR *p
 
 	from := state.CurrentMessageState.Contact.ID
 
+	// Backwards compatibility: old versions will not have the string Emoji, convert it
+	if pbEmojiR.Emoji == "" {
+		pbEmojiR.Emoji = ConvertEmojiIDToString(pbEmojiR.Type)
+	}
+
```
Older versions will not have this code 😄 While the newer versions will simply support the new `Emoji` property.

I think we can't implement it on reception. But we can support it on sender side by converting to the old emoji `Type` property.

jrainville replied:
This reception code makes it possible to still receive reactions from old versions.

We could do it on the sender side too, but I feel like being forward compatible is not as useful. Because let's say you are on 2.36, the reactions you send would only be shown to 2.35 and lower if they are exactly one of the six that we used to support :smile: 

So IMO, it's fine to no longer send reactions to older versions

igor-sirotin replied:
Oh indeed, it will work the other way around. Makes sense 👍

igor-sirotin commented on services/ext/api.go:
```diff
@@ -1020,7 +1020,12 @@ func (api *PublicAPI) RegisteredForPushNotifications() (bool, error) {
+}
+
+// TODO remove SendEmojiReaction and remove the 2 when the client supports sending custom emojis
+func (api *PublicAPI) SendEmojiReaction2(ctx context.Context, chatID, messageID string, emoji string) (*protocol.MessengerResponse, error) {
```
```suggestion
func (api *PublicAPI) SendEmojiReactionV2(ctx context.Context, chatID, messageID string, emoji string) (*protocol.MessengerResponse, error) {
```

igor-sirotin commented on protocol/protobuf/emoji_reaction.proto:
I wonder of a few things:
1. Later, can this be extended to custom emojis somehow?
I don't know myself. Just would be nice to avoid adding unnecessary obstacles for the future.
2. Maybe worth adding some validation to the emoji string? Otherwise it's an arbitrary string (not sure of the consequences).

jrainville replied:
> Later, can this be extended to custom emojis somehow?
I don't know myself. Just would be nice to avoid adding unnecessary obstacles for the future.

Good question. I know the idea was suggested as a feature request before. I don't know how we would do it though. 
I'll give it some thought and amend the proto if necessary

> Maybe worth adding some validation to the emoji string? Otherwise it's an arbitrary string (not sure of the consequences).

I added some validation here: https://github.com/status-im/status-go/pull/6916/files#diff-c6666d0e151b3306dedc920becb10a320d306699007eab9d68ddcbc668fc3d2eR55

Should it be in the proto?

igor-sirotin replied:
> I added some validation here

Pardon, missed it. Nothing needed in proto I think, the thing you have looks enough 👍

igor-sirotin reviewed (commented):
Can we please have a functional test for this?
We already have some `sendEmojiReaction` tests, they will need to be ammended at some point

igor-sirotin commented on protocol/messenger_emoji_reactions.go:
```diff
@@ -20,12 +42,28 @@ func (m *Messenger) SendEmojiReaction(ctx context.Context, chatID, messageID str
+	// Validate that the emoji is valid
+	// There is no foolproof way to validate emojis, but this regex should cover all standard emojis
+	// it will also allow some non-emoji unicode characters, but that's not a big issue
+	emojiRegex := regexp.MustCompile("(?:\u00A9|\u00AE|[\u2000-\u3300]|[\U0001F000-\U0001FBFF])")
```
Please extract this `regex` as a global variable in the beginning of the file. So that we don't compile it every time the we send an emoji. 

Perhaps also makes sense to validate incoming emojis?

jrainville committed:
chore: code review fixes

osmaczko commented on protocol/messenger_emoji_reactions.go:
```diff
@@ -11,7 +11,28 @@ import (
+
+}
+
+// TODO remove emojiID once the client supports sending custom emojis
```
Once this is done, do we plan to map emoji string to type so that old clients can see at least elementary reactions?

jrainville replied:
> Once this is done, do we plan to map emoji string to type so that old clients can see at least elementary reactions?

No, I don't think it's worth it. It's simple to map from old to new, but from new to old is a bit more arbitrary and also it's an incentive for people to upgrade :smile:

igor-sirotin reviewed (approved):
I wish to see a functional test for this ofc 😄 
Maybe @fbarbu15 can assist please?

jrainville commented:
> I wish to see a functional test for this ofc 😄 Maybe @fbarbu15 can assist please?

Good idea. I'll create an issue in the epic!

edit: https://github.com/status-im/status-desktop/issues/18822
